"""
gRPCè¿æ¥æµ‹è¯• - å¤šç§çŸ­ä¿¡å†…å®¹æµ‹è¯•
é€šè¿‡gRPCè¿æ¥SMSæœåŠ¡ï¼Œæµ‹è¯•å¤šç§çŸ­ä¿¡å†…å®¹
"""
import asyncio
import json
import time
import sys
import grpc
from pathlib import Path
from loguru import logger

# æ·»åŠ é¡¹ç›®æ ¹ç›®å½•åˆ°Pythonè·¯å¾„
project_root = Path(__file__).parent.parent
sys.path.insert(0, str(project_root))


# å°è¯•å¯¼å…¥gRPCæ¨¡å—
try:
    from src.sms_service import sms_pb2, sms_pb2_grpc
    HAS_GRPC = True
except ImportError as e:
    logger.error(f"âŒ æ— æ³•å¯¼å…¥gRPCæ¨¡å—: {e}")
    logger.error("è¯·å…ˆè¿è¡Œä»¥ä¸‹å‘½ä»¤ç”ŸæˆgRPCä»£ç :")
    logger.error("python -m grpc_tools.protoc -Iproto --python_out=src/sms_service --grpc_python_out=src/sms_service proto/sms.proto")
    sys.exit(1)


class GRPCSMSTester:
    """gRPC SMSæµ‹è¯•å™¨"""

    def __init__(self, target: str = "localhost:50052"):
        self.target = target
        self.channel = None
        self.stub = None

    async def connect(self) -> bool:
        """è¿æ¥åˆ°gRPCæœåŠ¡"""
        try:
            logger.info(f"ğŸ”— è¿æ¥åˆ°gRPCæœåŠ¡: {self.target}")
            self.channel = grpc.aio.insecure_channel(self.target)
            self.stub = sms_pb2_grpc.SMSServiceStub(self.channel)

            # æµ‹è¯•è¿æ¥
            try:
                await asyncio.wait_for(
                    self.stub.HealthCheck(sms_pb2.HealthCheckRequest()),
                    timeout=5.0
                )
                logger.info("âœ… gRPCè¿æ¥æˆåŠŸ")
                return True
            except asyncio.TimeoutError:
                logger.error("â° è¿æ¥è¶…æ—¶")
                return False
            except Exception as e:
                logger.error(f"âŒ è¿æ¥æµ‹è¯•å¤±è´¥: {e}")
                return False

        except Exception as e:
            logger.error(f"ğŸ’¥ è¿æ¥gRPCæœåŠ¡å¤±è´¥: {e}")
            return False

    async def close(self):
        """å…³é—­è¿æ¥"""
        if self.channel:
            await self.channel.close()
            logger.info("ğŸ”Œ å·²å…³é—­gRPCè¿æ¥")

    async def health_check(self):
        """å¥åº·æ£€æŸ¥"""
        try:
            logger.info("ğŸ©º æ‰§è¡Œå¥åº·æ£€æŸ¥...")
            response = await self.stub.HealthCheck(sms_pb2.HealthCheckRequest())

            if response.status == 200:
                logger.info(f"âœ… æœåŠ¡å¥åº·: {response.message}")
                try:
                    data = json.loads(response.data)
                    logger.info(f"   è°ƒåˆ¶è§£è°ƒå™¨: {data.get('modem_port', 'N/A')}")
                    logger.info(f"   ä¿¡å·å¼ºåº¦: {data.get('signal_strength', 'N/A')}")
                    logger.info(f"   è¿æ¥çŠ¶æ€: {data.get('modem_connected', 'N/A')}")
                except:
                    logger.info(f"   åŸå§‹æ•°æ®: {response.data[:100]}...")
                return True
            else:
                logger.error(f"âŒ æœåŠ¡ä¸å¥åº·: {response.message}")
                return False

        except Exception as e:
            logger.error(f"ğŸ’¥ å¥åº·æ£€æŸ¥å¤±è´¥: {e}")
            return False

    async def send_test_sms(self, test_case: dict) -> bool:
        """å‘é€æµ‹è¯•çŸ­ä¿¡"""
        try:
            logger.info(f"\nğŸ“‹ æµ‹è¯•ç”¨ä¾‹: {test_case['name']}")
            logger.info(f"ğŸ“± å‘é€åˆ°: {test_case['phone']}")
            logger.info(f"ğŸ“„ å†…å®¹é¢„è§ˆ: {test_case['content'][:80]}...")
            logger.info(f"ğŸ“ å†…å®¹é•¿åº¦: {len(test_case['content'])} å­—ç¬¦")

            # åˆ›å»ºè¯·æ±‚
            request = sms_pb2.SendSMSRequest(
                phone_number=test_case['phone'],
                content=test_case['content'],
                sender_id=f"test_{test_case['strategy']}",
                delivery_report=True,
                metadata={
                    "test_case": test_case['name'],
                    "strategy": test_case['strategy'],
                    "timestamp": str(time.time()),
                    "length": str(len(test_case['content']))
                }
            )

            # å‘é€çŸ­ä¿¡
            logger.info("ğŸš€ å‘é€çŸ­ä¿¡ä¸­...")
            start_time = time.time()
            response = await self.stub.SendSMS(request)
            elapsed_time = time.time() - start_time

            if response.status == 200:
                logger.info(f"âœ… å‘é€æˆåŠŸ ({elapsed_time:.2f}s)")
                logger.info(f"   çŠ¶æ€: {response.message}")
                try:
                    data = json.loads(response.data)
                    logger.info(f"   æ¶ˆæ¯ID: {data.get('message_id', 'N/A')}")
                    logger.info(f"   å‚è€ƒå·: {data.get('reference', 'N/A')}")
                    logger.info(f"   è°ƒåˆ¶è§£è°ƒå™¨ç«¯å£: {data.get('modem_port', 'N/A')}")
                except:
                    logger.info(f"   æ•°æ®: {response.data[:100]}...")
                return True
            else:
                logger.error(f"âŒ å‘é€å¤±è´¥ ({elapsed_time:.2f}s)")
                logger.error(f"   é”™è¯¯: {response.message}")
                logger.error(f"   çŠ¶æ€ç : {response.status}")
                return False

        except Exception as e:
            logger.error(f"ğŸ’¥ å‘é€è¿‡ç¨‹ä¸­å‡ºé”™: {e}")
            return False


async def main():
    """ä¸»æµ‹è¯•å‡½æ•°"""
    import argparse

    parser = argparse.ArgumentParser(description="gRPC SMSå¤šå†…å®¹æµ‹è¯•")
    parser.add_argument("--target", "-t", default="localhost:50052",
                       help="gRPCæœåŠ¡å™¨åœ°å€")
    parser.add_argument("--phone", "-p", default="+8619834717434",
                       help="æµ‹è¯•æ‰‹æœºå·ç ")

    args = parser.parse_args()

    # é…ç½®æ—¥å¿—
    logger.remove()
    logger.add(
        lambda msg: print(msg, end=""),
        format="<green>{time:YYYY-MM-DD HH:mm:ss}</green> | "
               "<level>{level: <8}</level> | "
               "<cyan>{name}</cyan>:<cyan>{function}</cyan> - "
               "<level>{message}</level>",
        level="INFO",
        colorize=True
    )

    # åˆ›å»ºæµ‹è¯•å™¨
    tester = GRPCSMSTester(args.target)

    try:
        # 1. è¿æ¥åˆ°gRPCæœåŠ¡
        logger.info("\n" + "="*60)
        logger.info("æ­¥éª¤1: è¿æ¥åˆ°gRPCæœåŠ¡")
        logger.info("="*60)

        if not await tester.connect():
            logger.error("âŒ æ— æ³•è¿æ¥åˆ°gRPCæœåŠ¡ï¼Œæµ‹è¯•ç»ˆæ­¢")
            return

        # 2. å¥åº·æ£€æŸ¥
        logger.info("\n" + "="*60)
        logger.info("æ­¥éª¤2: æœåŠ¡å¥åº·æ£€æŸ¥")
        logger.info("="*60)

        if not await tester.health_check():
            logger.warning("âš ï¸  æœåŠ¡å¥åº·æ£€æŸ¥å¤±è´¥ï¼Œä½†ç»§ç»­æµ‹è¯•...")

        # 3. å‡†å¤‡æµ‹è¯•ç”¨ä¾‹ï¼ˆä½¿ç”¨ä¸å®¢æˆ·ç«¯ç›¸åŒçš„å†…å®¹æ ¼å¼ï¼‰
        test_cases = [
            {
                "name": "çº¯è‹±æ–‡çŸ­ä¿¡",
                "strategy": "english_only",
                "phone": args.phone,
                "content": f"Hello World! Test SMS from Python gRPC client at {time.strftime('%H:%M:%S')}. This is a pure English test message to verify the SMS service functionality."
            },
            {
                "name": "çº¯ä¸­æ–‡çŸ­ä¿¡",
                "strategy": "chinese_only",
                "phone": args.phone,
                "content": f"æµ‹è¯•çŸ­ä¿¡ {time.strftime('%Y-%m-%d %H:%M:%S')} - è¿™æ˜¯ä¸€æ¡çº¯ä¸­æ–‡æµ‹è¯•çŸ­ä¿¡ï¼Œç”¨äºéªŒè¯Python SMSå¾®æœåŠ¡çš„gRPCæ¥å£åŠŸèƒ½ã€‚çŸ­ä¿¡æœåŠ¡åº”è¯¥æ­£ç¡®å¤„ç†ä¸­æ–‡ç¼–ç ã€‚"
            },
            {
                "name": "ä¸­è‹±æ–‡æ··åˆçŸ­ä¿¡",
                "strategy": "mixed_content",
                "phone": args.phone,
                "content": f"æµ‹è¯•çŸ­ä¿¡ Test Message {time.strftime('%H:%M:%S')} - è¿™æ˜¯ä¸€æ¡ä¸­è‹±æ–‡æ··åˆçš„æµ‹è¯•çŸ­ä¿¡ Mixed Chinese and English test message.\n\n"
                          f"English part: This SMS is sent via gRPC interface.\n"
                          f"ä¸­æ–‡éƒ¨åˆ†ï¼šè¿™æ¡çŸ­ä¿¡é€šè¿‡gRPCæ¥å£å‘é€ã€‚\n"
                          f"æ··åˆæµ‹è¯•ï¼šTest æµ‹è¯• Mixed æ··åˆ Message æ¶ˆæ¯"
            },
            {
                "name": "å¸¦ç‰¹æ®Šç¬¦å·çš„çŸ­ä¿¡",
                "strategy": "special_chars",
                "phone": args.phone,
                "content": f"ã€é‡è¦é€šçŸ¥ã€‘Test SMS {time.strftime('%Y-%m-%d')} 14:30\n"
                          f"ğŸ“± æµ‹è¯•å†…å®¹åŒ…å«ç‰¹æ®Šç¬¦å·ï¼š\n"
                          f"â€¢ é¡¹ç›®ç¬¦å·\n"
                          f"â˜… æ˜Ÿå·ç¬¦å·\n"
                          f"â†’ ç®­å¤´ç¬¦å·\n"
                          f"ğŸ’° è´§å¸ç¬¦å·\n"
                          f"âœ… å®Œæˆç¬¦å·\n\n"
                          f"Special chars: !@#$%^&*()_+-=[]{{}}|;:',.<>/?~`"
            },
            {
                "name": "é•¿æ–‡æœ¬çŸ­ä¿¡æµ‹è¯•",
                "strategy": "long_text",
                "phone": args.phone,
                "content": f"é•¿æ–‡æœ¬çŸ­ä¿¡æµ‹è¯• {time.strftime('%H:%M:%S')}\n\n"
                          f"è¿™æ˜¯ä¸€æ¡è¾ƒé•¿çš„æµ‹è¯•çŸ­ä¿¡ï¼Œç”¨äºæµ‹è¯•çŸ­ä¿¡æœåŠ¡çš„æ–‡æœ¬å¤„ç†èƒ½åŠ›ã€‚"
                          f"çŸ­ä¿¡å†…å®¹åŒ…å«å¤šä¸ªæ®µè½å’Œæ¢è¡Œç¬¦ï¼Œç¡®ä¿æœåŠ¡èƒ½å¤Ÿæ­£ç¡®å¤„ç†ã€‚\n\n"
                          f"English paragraph: This is a longer test message to verify the SMS service's "
                          f"ability to handle extended text content with mixed Chinese and English characters. "
                          f"The message includes multiple paragraphs and line breaks.\n\n"
                          f"æœ€åä¸€æ®µï¼šæµ‹è¯•ç»“æŸã€‚æ„Ÿè°¢ä½¿ç”¨Python SMSå¾®æœåŠ¡ã€‚"
                          f"å‚è€ƒç¼–å·ï¼šTEST-{int(time.time())}"
            },
            {
                "name": "å®¢æˆ·ç«¯å®é™…å†…å®¹æµ‹è¯•",
                "strategy": "client_actual",
                "phone": args.phone,
                "content": f"æµ‹è¯•çŸ­ä¿¡ {time.strftime('%Y-%m-%d %H:%M:%S')} - è¿™æ˜¯ä¸€æ¡æ¥è‡ªPython SMSå¾®æœåŠ¡çš„æµ‹è¯•çŸ­ä¿¡ã€‚\n" +
                          f"{'-'*15}\n" +
                          f"Test SMS {time.strftime('%Y-%m-%d %H:%M:%S')} - This is a test from Python SMS service."
            }
        ]

        # 4. æ‰§è¡Œæµ‹è¯•ç”¨ä¾‹
        logger.info("\n" + "="*60)
        logger.info("æ­¥éª¤3: æ‰§è¡Œå¤šç§çŸ­ä¿¡å†…å®¹æµ‹è¯•")
        logger.info("="*60)

        total_cases = len(test_cases)
        success_count = 0
        failure_count = 0

        for i, test_case in enumerate(test_cases):
            logger.info(f"\nğŸ“Š æµ‹è¯•è¿›åº¦: {i+1}/{total_cases}")
            logger.info("-" * 40)

            success = await tester.send_test_sms(test_case)

            if success:
                success_count += 1
            else:
                failure_count += 1

            # ç­‰å¾…ä¸€ä¸‹å†å‘é€ä¸‹ä¸€æ¡ï¼ˆé¿å…å¤ªå¿«ï¼‰
            if i < total_cases - 1:
                logger.info(f"â³ ç­‰å¾…3ç§’åç»§ç»­...")
                await asyncio.sleep(3)

        # 5. æµ‹è¯•ç»“æœæ±‡æ€»
        logger.info("\n" + "="*60)
        logger.info("æµ‹è¯•ç»“æœæ±‡æ€»")
        logger.info("="*60)

        logger.info(f"ğŸ“ˆ æ€»æµ‹è¯•ç”¨ä¾‹: {total_cases}")
        logger.info(f"âœ… æˆåŠŸ: {success_count}")
        logger.info(f"âŒ å¤±è´¥: {failure_count}")
        logger.info(f"ğŸ“Š æˆåŠŸç‡: {success_count/total_cases*100:.1f}%")

        if success_count == total_cases:
            logger.info("ğŸ‰ æ‰€æœ‰æµ‹è¯•ç”¨ä¾‹éƒ½é€šè¿‡äº†ï¼gRPC SMSæœåŠ¡è¿è¡Œæ­£å¸¸ï¼")
        elif success_count > 0:
            logger.info(f"âš ï¸  éƒ¨åˆ†æµ‹è¯•é€šè¿‡ ({success_count}/{total_cases})")
        else:
            logger.error("ğŸ’¥ æ‰€æœ‰æµ‹è¯•éƒ½å¤±è´¥äº†ï¼è¯·æ£€æŸ¥SMSæœåŠ¡é…ç½®å’Œè°ƒåˆ¶è§£è°ƒå™¨è¿æ¥ã€‚")

        # 6. æ‰¹é‡å‘é€æµ‹è¯•ï¼ˆå¯é€‰ï¼‰
        logger.info("\n" + "="*60)
        logger.info("æ­¥éª¤4: æ‰¹é‡å‘é€æµ‹è¯•ï¼ˆå¯é€‰ï¼‰")
        logger.info("="*60)

        try:
            batch_request = sms_pb2.SendBatchSMSRequest(
                phone_numbers=[args.phone, args.phone],  # åŒä¸€ä¸ªå·ç ä¸¤æ¬¡ï¼Œæµ‹è¯•æ‰¹é‡åŠŸèƒ½
                content=f"æ‰¹é‡æµ‹è¯•çŸ­ä¿¡ {time.strftime('%H:%M:%S')} - è¿™æ˜¯æ‰¹é‡åŠŸèƒ½æµ‹è¯•",
                sender_id="batch_test",
                delivery_report=False,
                metadata={"batch_test": "true", "timestamp": str(time.time())}
            )

            logger.info("ğŸš€ å‘é€æ‰¹é‡çŸ­ä¿¡æµ‹è¯•...")
            batch_response = await tester.stub.SendBatchSMS(batch_request)

            if batch_response.status == 200:
                logger.info(f"âœ… æ‰¹é‡å‘é€æˆåŠŸ: {batch_response.message}")
                try:
                    batch_data = json.loads(batch_response.data)
                    logger.info(f"   æ‰¹æ¬¡ID: {batch_data.get('batch_id', 'N/A')}")
                    logger.info(f"   æ€»æ•°é‡: {batch_data.get('total_count', 0)}")
                    logger.info(f"   æˆåŠŸ: {batch_data.get('success_count', 0)}")
                    logger.info(f"   å¤±è´¥: {batch_data.get('failed_count', 0)}")
                except:
                    pass
            else:
                logger.warning(f"âš ï¸  æ‰¹é‡å‘é€å¤±è´¥: {batch_response.message}")

        except Exception as e:
            logger.warning(f"âš ï¸  æ‰¹é‡å‘é€æµ‹è¯•è·³è¿‡: {e}")

        logger.info("\n" + "="*60)
        logger.info("ğŸ¯ æµ‹è¯•å®Œæˆ")
        logger.info("="*60)

    except KeyboardInterrupt:
        logger.info("\nâŒ¨ï¸ ç”¨æˆ·ä¸­æ–­æµ‹è¯•")
    except Exception as e:
        logger.error(f"ğŸ’¥ æµ‹è¯•è¿‡ç¨‹ä¸­å‘ç”Ÿé”™è¯¯: {e}")
        import traceback
        traceback.print_exc()
    finally:
        await tester.close()


if __name__ == "__main__":
    asyncio.run(main())
